daemon/c_run.c

```diff
@@ -0,0 +1,146 @@
+#define _GNU_SOURCE
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <sys/wait.h>
+#include <sys/syscall.h>
+
+#define LOGF_LOG_MIN_PRIO LOGF_PRIO_TRACE
+#include <macro.h>
+
+#include "common/mem.h"
+#include "common/dir.h"
+#include "common/file.h"
+#include "hardware.h"
+#include "container.h"
+#include "c_run.h"
+#include "c_cgroups.h"
+#include "c_cap.h"
+
+int setns(int fd, int nstype)
+{
+	return syscall(__NR_setns, fd, nstype);
+}
+
+#define MAX_NS 6
+int fd[MAX_NS] = { 0 };
+
+static int
+setns_cb(const char *path, const char *file, void *data)
+{
+	int *i = data;
+
+	char *ns_file = mem_printf("%s%s", path, file);
+	TRACE("Opening namespace file %s", ns_file);
+
+	if (*i >= MAX_NS) {
+		ERROR("Too many namespace files found in %s", path);
+		goto error;
+	}
+
+	fd[*i] = open(ns_file, O_RDONLY);
+	if (fd[*i] == -1) {
+		ERROR_ERRNO("Could not open namespace file %s", ns_file);
+		goto error;
+	}
+
+	*i = *i + 1;
+
+	mem_free(ns_file);
+	return 0;
+
+error:
+	mem_free(ns_file);
+	abort();
+}
+
+int
+c_run_set_namespaces(pid_t pid)
+{
+	char *pid_string = mem_printf("%d", pid);
+
+	TRACE("Setting namespaces for pid %s", pid_string);
+
+	// set namespaces
+	char *folder = mem_printf("/proc/%d/ns/", pid);
+
+	int i = 0;
+	if (dir_foreach(folder, &setns_cb, &i)) {
+		FATAL("Could not traverse PID dir in procfs, wrong PID?");
+	}
+
+	for (int j = 0; j < i; j++) {
+		if (setns(fd[j], 0) == -1) {	/* Join that namespace */
+			FATAL_ERRNO("Could not join namespace");
+		}
+	}
+
+	TRACE("Successfully joined all namespaces");
+
+	return 0;
+}
+
+int
+c_run_set_cgroups(const container_t *container, const pid_t pid)
+{
```

Comment:
why parameter pid, container struct traks pid of its init proces: use container_get_pid()